walkeeper/src/timeline.rs

```diff
@@ -150,80 +150,114 @@ impl SharedState {
-            Ok(mut file) => {
-                // Lock file to prevent two or more active safekeepers
-                match file.try_lock_exclusive() {
-                    Ok(()) => {}
-                    Err(e) => {
-                        bail!(
-                            "control file {:?} is locked by some other process: {}",
-                            &control_file_path,
-                            e
-                        );
-                    }
-                }
-                // Empty file is legit on 'create', don't try to deser from it.
-                if file.metadata().unwrap().len() == 0 {
-                    if let CreateControlFile::False = create {
-                        bail!("control file is empty");
-                    }
-                    Ok((file, SafeKeeperState::new()))
-                } else {
-                    match SafeKeeperState::des_from(&mut file) {
-                        Err(e) => {
-                            bail!("failed to read control file {:?}: {}", control_file_path, e);
-                        }
-                        Ok(s) => {
-                            if s.magic != SK_MAGIC {
-                                bail!("bad control file magic: {}", s.magic);
-                            }
-                            if s.format_version != SK_FORMAT_VERSION {
-                                bail!(
-                                    "incompatible format version: {} vs. {}",
-                                    s.format_version,
-                                    SK_FORMAT_VERSION
-                                );
-                            }
-                            Ok((file, s))
-                        }
-                    }
-                }
+
+        let lock_file = File::create(&lock_file_path).with_context(|| "failed to open lockfile")?;
+
+        // Lock file to prevent two or more active safekeepers
+        lock_file.try_lock_exclusive().map_err(|e| {
+            anyhow!(
+                "control file {:?} is locked by some other process: {}",
+                &control_file_path,
+                e
+            )
+        })?;
+
+        let mut control_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(matches!(create, CreateControlFile::True))
+            .open(&control_file_path)
+            .with_context(|| {
+                format!(
+                    "failed to open control file at {}",
+                    control_file_path.display(),
+                )
+            })?;
+
+        // Empty file is legit on 'create', don't try to deser from it.
+        let state = if control_file.metadata().unwrap().len() == 0 {
+            if let CreateControlFile::False = create {
+                bail!("control file is empty");
+            }
+            SafeKeeperState::new()
+        } else {
+            let mut buf = Vec::new();
+            control_file
+                .read_to_end(&mut buf)
```

Comment:
Nit: I'd `ensure!` that file is not too big before reading. Say, ten megabytes. Maybe log an error if it's greater than 5 MB. Should never happen, but who knows.

Reply:
I think it's ok to just issue a warning after the read if its too large. If we care about it's size we should check it on write too. @arssher thoughts?

Reply:
I'm mostly worried about accidentally **ram**ming 5 tb of data. Unconstrained reads scare me a little.

Reply:
If this happens it is a consequence, not the root cause. If we intent to guard against accidently large safekeeper state we should prevent it from growing that big. Otherwise you see normally working safekeeper which fails if you restart it.

Reply:
True, but this file is out of our control while it lies on the disk. Random `cp some-big-data-file control-file` would be bad. Or if FS is corrupted and thinks the file is much bigger than it is.

Of course, should never happen, and I don't have experience with such scenarios, but still.

Reply:
Also note that we rewrite it on every `persist` call

Reply:
I'm not too worried about this, the probability is small and the consequences are not fatal. However, note that if you put checksum in the beginning of the file, you can feed fd directly to `des_from` which will bail out immediately once file structure doesn't match.